### ISSUE TITLE
simple_cycles infinite loop

### ISSUE THREAD

**Issue by steven-nicolaou** [networkx/networkx#6052]
### Current Behavior
`simple_cycles` appears to enter an infinite loop. My graph has under 4000 nodes and edges. If I subtract a handful of them (less than 100), it runs successfully in just about a minute. I could not reduce it to a minimal example, so I have hashed the nodes and attached the smallest example that I could reproduce it with. It does not seem to be just a matter of a large number of cycles.

### Expected Behavior
`simple_cycles` returns a list of cycles

### Steps to Reproduce
g1: Graph that triggers infinite loop
g2: Graph with some edges removed that runs successfully

```
# Load Graph 1
g1 = nx.adjacency_graph(json.load(open('g1.json')))
cycles = nx.simple_cycles(g1) # ← Infinite loop

# Load Graph 2
g2 = nx.adjacency_graph(json.load(open('g2.json')))
cycles = nx.simple_cycles(g2) # ← Runs in about 1 minute
```

### Environment
Python version: 3.9.12
NetworkX version: 2.8.7

### Additional context
The 2 graphs in JSON format
[graphs.zip](https://github.com/networkx/networkx/files/9770299/graphs.zip)

**Comment by paulitapb** [comment]
Hi, I could replicate your problem. How did you select which nodes to subtract?

**Comment by stevennic** [comment]
Just trial and error, but it seems to happen with arbitrary combinations of nodes, so I don't think the problem is with individual nodes per se. This is just as minimal as I could get it.

**Comment by paulitapb** [comment]
ok! I was looking at the algorithm cover test and there's one if-condition branch that was not tested. I will see if that's where the problem occurs.

**Comment by paulitapb** [comment]
good point! I think that may be the problem. Because for g2 the number of simple cycles is 89113 which is not much compared with the worst case for a graph of 3587 nodes. And as @steven-nicolaou chose the number of nodes and edges randomly it's very likely that g1 is a graph with too many simple cycles.

**Comment by steven-nicolaou** [comment]
Interesting point and I was afraid about the possibility of 1 edge adding an exponential number of cycles. But it also crashes due to out of memory error if I let it run long enough. I thought it might be a memory leak, but if it's something that simply needs a long time to run, it should do so without crashing.

**Comment by rfulekjames** [comment]
I haven't observed any memory leak while enumerating cycles of g1 for several hours. Since the function is a generator, storing simple cycles in memory throughout its execution is not necessary. Therefore I am not sure where the problem could be.

**Comment by dschult** [comment]
If you try to store all the simple cycles in a e.g. list, you will run out of memory. Also, the code you posted above can't be the code you actually use because it only creates the generator -- and doesn't iterate over it -- so no work is actually done.  Are you using it with code that will store all the paths?

**Comment by antoscha** [comment]
The function is still not working

**Comment by dschult** [comment]
Do you mean this function is still working?   :)  The issue is that the function never stopped. So I guessit is still working. :)

@antoscha, If you want us to work on this, then you will need to provide more information. The current OP code runs instantly because it only creates the iterator. It doesn't iterate over the iterator. And the comments suggest that 
1) the number of cycles can grow very rapidly as some edges are added (and with other edges they grow very slowly). 
2) this explosion in number of cycles could fill memory if the iterator output is stored -- that could explain using up memory.

What do you have that is new information about this fuction?

**Comment by antoscha** [comment]
The information I have is the function apparently generates one concrete cycle infinitely. A simple workaround I used for a directed graph:
```
def find_cycles(graph: DiGraph) -> list[tuple[str, ...]]:
    cycles: set[tuple[str, ...]] = set()

    def loop(node: str, path: tuple[str, ...]) -> None:
        visited.add(node)
        for successor in graph.successors(node):
            if successor not in visited:
                loop(successor, (*path, successor))
            elif successor in path:
                cycles.add(path[path.index(successor):])

    for node in graph.nodes:
        visited: set[str] = set()
        loop(node, tuple(node))

    return list(cycles)
```

**Comment by boothby** [comment]
Would you please provide a graph that exhibits that behavior?

**Comment by antoscha** [comment]
Sorry, it's a proprietary thing.

**Comment by boothby** [comment]
I have managed to extract minimal working examples from proprietary data before.  Does the algorithm exhibit the same behavior on the induced graph of the single cycle you're seeing repeated?  That is, if `C` is the first cycle and you run `simple_cycles(G.subgraph(C))` does it still repeat it?  If that's the case, can you remove the labels from that graph with `nx.relabel.convert_node_labels_to_integers`?  That is typically enough to strip any proprietary information.

Your implementation is incorrect, unfortunately.  In addition to reporting the same cycle multiple times (from different starting points) it fails to list some cycles due to how you're using `visited`.

```python
D = nx.DiGraph([
  (0, 1), (0, 2), (0, 3),
  (1, 2), (1, 3), (2, 3),
  (1, 4), (2, 4), (3, 4),
  (4, 0)
])
```

When I run your code on the graph above, it does not find the cycle `(0, 2, 4)`.

Note that your use of `visited` only allows each node to be visited a single time during the exploration rooted at each given node.  Therefore, only a quadratic number of cycles can be found.  Where this is guaranteed to be fast, it must be incorrect as there may be an exponential number of simple cycles in a graph, for example using the graph `D` below with even `n`, the number of cycles grows like the fibonacci numbers (see https://oeis.org/A005248 for the counting sequence)

```python
D = nx.cycle_graph(range(n), create_using=nx.DiGraph)
nx.add_cycle(D, range(0, n, 2))
nx.add_cycle(D, range(1, n, 2))
```

**Comment by phschaad** [comment]
I am still observing this issue in networkx version 3.4.2, has anyone looked into this since 2023?

**Comment by dschult** [comment]
What part of this issue are you observing?

If you want us to work on this, then you will need to provide more information. The current OP code runs instantly because it only creates the iterator. It doesn't iterate over the iterator. And the comments suggest that

- the number of cycles can grow very rapidly as some edges are added (and with other edges they grow very slowly).
- this explosion in number of cycles could fill memory if the iterator output is stored -- that could explain using up memory.

The example above from @antoscha uses a function that has `return list(cycles)`.  That is likely to cause memory problems because it is trying to store all the cycles in memory.  If your problem is a memory issue, try running `sum(1 for _ in nx.simple_cycles(G)` to see if that finishes... Or you could write the results to a file as they are created -- anything to keep them all from being stored in memory.

What do you have that is new information about this function?

**Comment by rossbar** [comment]
Reading through this thread, I think there are at least 2 (perhaps more) independent discussions happening:
 1. The issue from the [original post](https://github.com/networkx/networkx/issues/6052#issue-1407107111) re: a suspected infinite loop
 2. A second issue related to [a single cycle being produced infinitely](https://github.com/networkx/networkx/issues/6052#issuecomment-1826799456)

I believe `1.` is not a bug and has been [sufficiently addressed](https://github.com/networkx/networkx/issues/6052#issuecomment-1284137961). The number of cycles in a graph can change drastically with even very small modifications to the edges. What seems like an "infinite loop" is much more likely to be due to the fact that there may be *many* more cycles in the graph than one might expect. For example, take a look at how the number of cycles grows in [undirected complete graphs](https://oeis.org/A002807). As a concrete reference point, running `sum(1 for _ in nx.simple_cycles(nx.complete_graph(11)))` takes about 6.5 seconds to count all 5,488,059 cycles on my laptop. If we assume that the computation time scales with the number of cycles (a best-case scenario) then I would expect the same computation for `complete_graph(15)` to take 23,000x longer, or about 42 hours. Not an *infinite* loop, but definitely longer than I'd be willing to wait!

The memory component of this comes from trying to store all the cycles --- in fact, this is one of the main reasons why `simple_cycles` is implemented as a generator! Sticking with `complete_graph(15)` as an example: there are `127661752406` cycles in this graph. For the sake of argument, let's assume the average cycle length is 3[^1] and each node is represented by a 64-bit integer. The total memory footprint for storing all the cycles[^2] would be `num_cycles * num_nodes_per_cycle * num_bytes_per_node` = `127661752406 * 3 * 8` = ~3 TB. In other words, the explosion in memory is a symptom of the same fundamental cause: there may be *many* cycles in the graph - so many that storing all of them in memory is not feasible.

Point `2.` sounds like it could be a real bug: `simple_cycles` is not expected to yield the same cycle over and over; however, absent a reproducing example there's no path forward to investigate whether the problem is related to `simple_cycles` or something else.

Given the above, I'm going to close this issue. For users who still have issues with `simple_cycles`, please open a new issue (preferrably with a minimal reproducing example) so that others can try to help concretely address it!

[^1]: We **know** it's longer because 3 is the minimum length of a cycle and there are certainly longer cycles in `K15`, but this is fine for an overly conservative estimate that illustrates the point.

[^2]: Ignoring overhead from data structures - just the data itself!
